experimental/skills/DialogToDialog/DialogRootBot/Dialogs/SkillDialog.cs

```diff
@@ -0,0 +1,153 @@
+        }
+
+        public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default)
+        {
+            if (options is CancellationToken)
+            {
+                throw new ArgumentException($"{nameof(options)} cannot be a cancellation token");
+            }
+
+            if (!(options is SkillDialogArgs dialogArgs))
+            {
+                throw new ArgumentNullException(nameof(options), $"Unable to cast {nameof(options)} to {nameof(SkillDialogArgs)}");
+            }
+
+            var skillId = dialogArgs.SkillId;
+            if (!_skillsConfig.Skills.TryGetValue(skillId, out var skillInfo))
+            {
+                throw new KeyNotFoundException($"Unable to find \"{skillId}\" in the skill configuration.");
+            }
+
+            // Store Skill ID for this dialog instance
+            await _activeSkillProperty.SetAsync(dc.Context, skillId, cancellationToken);
+
+            await dc.Context.SendTraceActivityAsync($"{GetType().Name}.BeginDialogAsync(). Using activity of type: {dialogArgs.ActivityType}", cancellationToken: cancellationToken);
+
+            Activity skillActivity;
+            switch (dialogArgs.ActivityType)
+            {
+                case ActivityTypes.Event:
+                    var eventActivity = Activity.CreateEventActivity();
+                    eventActivity.Name = dialogArgs.Name;
+                    eventActivity.ApplyConversationReference(dc.Context.Activity.GetConversationReference(), true);
+                    skillActivity = (Activity)eventActivity;
+                    break;
+
+                case ActivityTypes.Message:
+                    var messageActivity = Activity.CreateMessageActivity();
+                    messageActivity.Text = dialogArgs.Text;
+                    skillActivity = (Activity)messageActivity;
+                    break;
+
+                default:
+                    throw new ArgumentException($"Invalid activity type in {dialogArgs.ActivityType} in {nameof(SkillDialogArgs)}");
+            }
+
+            ApplyParentActivityProperties(dc, skillActivity, dialogArgs);
+            return await SendToSkill(dc, skillActivity, skillInfo, cancellationToken);
+        }
+
+        public override async Task<DialogTurnResult> ContinueDialogAsync(DialogContext dc, CancellationToken cancellationToken = default)
+        {
+            await dc.Context.SendTraceActivityAsync($"{GetType().Name}.ContinueDialogAsync(). ActivityType: {dc.Context.Activity.Type}", cancellationToken: cancellationToken);
+
+            var skillId = await _activeSkillProperty.GetAsync(dc.Context, () => null, cancellationToken);
+
+            if (dc.Context.Activity.Type == ActivityTypes.Message && dc.Context.Activity.Text.Equals("abort", StringComparison.CurrentCultureIgnoreCase))
```

Comment:
Should we use a pattern similar to the Core bot's [cancel and help dialog](https://github.com/microsoft/BotBuilder-Samples/blob/master/samples/csharp_dotnetcore/13.core-bot/Dialogs/CancelAndHelpDialog.cs)?

Reply:
Maybe, let me think about this one, I don't want to create a new base class if possible.

---
In reply to: [356232091](https://github.com/microsoft/BotBuilder-Samples/pull/2032#discussion_r356232091) [](ancestors = 356232091)